contributors/design-proposals/raw-block-pv.md

```diff
@@ -0,0 +1,531 @@
+  * An admin wishes to expose a block volume to be consumed as a block volume for the user  
+  * An admin wishes to expose a block volume to be consumed as a block volume for an administrative function such 
+    as bootstrapping 
+  * A user wishes to utilize initiate block storage to fully realize the performance of an application tuned to using block devices
+  Future use cases include dynamically provisioning and intelligent discovery of existing devices, which this proposal sets the 
+  foundation for more fully developing these methods. 
+  
+  It is important to note that when a PV is bound, it is either bound as a raw block device or formatted with a filesystem. Therefore, 
+  the PVC drives the request and intended usage of the device by specifying the volumeMode as part of the API. This design lends itself
+  to future support of dynamic provisioning by also letting the request initiate from the PVC defining the role for the PV. It also 
+  allows flexibility in the implementation and storage plugins to determine their support of this feature.
+ 
+# Design Overview
+
+  The proposed design is based on the idea of leveraging well defined concepts for storage in Kubernetes. The consumption and 
+  definitions for the block devices will be driven through the PVC and PV definitions. Along with Storage
+  Resource definitions, this will provide the admin with a consistent way of managing all storage. 
+  The API changes proposed in the following section are minimal with the idea of defining a volumeMode to indicate both the definition
+  and consumption of the devices. Since it's possible to create a volume as a block device and then later consume it by provisioning
+  a filesystem on top, the design requires explicit intent for how the volume will be used.
+  The additional benefit of explicitly defining how the volume is to be consumed will provide a means for indicating the method
+  by which the device should be scrubbed when the claim is deleted, as this method will differ from a raw block device compared to a 
+  filesystem. The ownership and responsibility of defining the retention policy shall be up to the plugin method being utilized and is
+  not covered in this proposal.
+  
+  Limiting use of the volumeMode to block can be executed through the use of storage resource quotas and storageClasses defined by the 
+  administrator.
+  
+  To ensure backwards compatibility and a phased transition of this feature, the consensus from the community is to intentionally disable
+  the volumeMode: Block for external provisioners until a suitable implementation for provisioner versioning has been accepted and 
+  implemented in the community. This requirement is better described in the design PR discussion and will be implemented as a separate
+  initiative. 
+  In addition, in-tree provisioners should be able to gracefully ignore volumeMode API objects for plugins that haven't been updated to
```

Comment:
I guess how it would work is that if the PVC requested block, the provisioner would provision a filesystem PV, and then we wouldn't bind it since the volumeModes don't match?

Reply:
@msau42  Correct. It isn't ideal and we will have to delete the PV that was created (hence the sub-proposal in this PR for the versioning) to get around this behavior.

Reply:
Who will do the deletion?  The pv controller or the user?  Should the code not invoke provisioning for PVC.volumeMode == block for this first phase?

Reply:
@msau42 pvcontroller would do the deletion (ideally). And yes, the code would invoke provisioning for PVC.volumeMode == block (however in phase one we aren't doing any dynamic provisioning)

Reply:
Since in the first phase, we're not doing dynamic provisioning, then the code needs to be updated to not invoke provisioning for in-tree provisioners too.

Reply:
@msau42 The design already states this "In addition, in-tree provisioners should be able to gracefully ignore volumeMode API objects for plugins that haven't been updated"

Reply:
That implies in the implementation, we either have to:
1) Update all in-tree provisioners to ignore the new PVC.volumeMode == block, or
2) Update the pv controller to not invoke dynamic provisioning for PVC.volumeMode == block.

I just want to make sure it's clear that even though we're not supporting dynamic provisioning of block PVs this phase, there are still code changes required to make sure we don't invoke provisioning.

Reply:
@msau42 I believe #2 would be the right approach. Thoughts?

Reply:
Agree, I think that's the simplest approach.

Reply:
@msau42 added note in the design overview